CddaSongLoader now reads tags from CD-Text

lumip opened the pull request:
### Changes so far
In `CddaSongLoader::LoadSongsFromCdda`: I changed the gstreamer source element to `cdiocddasrc` (from uri source), which includes metadata from CD-Text in the tag message for each track and seek through all the tracks. On a high level `LoadSongsFromCdda` now works as follows:

1. read the number of tracks, emit `SongsLoaded` (unchanged)
2. handle TOC message, update song durations, emit `SongDurationLoaded` (unchanged)
3. seek through tracks, fetch CD-Text metadata, emit `SongMetadataLoaded` (new)
4. emit `MusicBrainzDiscIdLoaded`
5. fetch metadata from musicbrainz and emit `SongMetadataLoaded` (as reaction to `MusicBrainzDiscIdLoaded`)

The benefits are that, if CD-Text is present on the disc, at least some metadata are still available even if the user does not have Internet connectivity (or musicbrainz service is unresponsive); it could also help disambiguating non-unique discids (see #7020 ), but that's not currently implemented.

In terms of clients of `CddaSongLoader`, I have adapted the `RipCDDialog` to only update metadata if it receives previously unseed tags, which means it currently fills with CD-Text tags first and then completes everything missing from the musicbrainz response. The only other client is `SongLoader`, which I have left unchanged so far because I wasn't sure of the exact implications of changing stuff there.

#### Other changes in `CddaSongLoader`:

- The `cdio_` object was opened but never used in `LoadSongsFromCdda`. I have removed it there and now initialise it during class construction, would need some input on preferred ways of error handling there (throw an exception?).
- This has removed the need for the mutex as cdio_ is only ever accessed from the main thread (and only via the `HasChanged`) method.
- There was a concurrency bug where the destructor might have been called while the thread running `LoadSongsFromCdda` was still active. I resolved that by storing the `QFuture` returned by `QtConcurrent::run` and waiting on it in the class destructor.

### Going further
I am thinking making the following further changes but would like some feedback on them first:
I dislike that `CddaSongLoader` now emits the `SongMetadataLoaded` signal two times containing potentially different metadata and it is up to the clients to deal with that. I'm thinking of either having `CddaSongLoader` maintain the `SongList` of CD tracks internally and update it when it gets input from a metadata source, then emit the `SongMetadataLoaded` signal ((maybe then rename that to `SongMetadataUpdated`) with a guarantee that previously emitted tags do not change. Clients could then always just use the latest SongList received, which would avoid duplicating code for dealing with this in the clients. While this would introduce more state to `CddaSongLoader`, I don't think there would be any harm in that. It could even be beneifical because the class could then encapsulate media changed checks, effectively only rerunning `LoadSongsFromCdda` if a media change was detected. This logic is currently pushed into its clients.

A stateless alternative would be to not have `CddaSongLoader` do a musicbrainz request and leave that to clients subscribing to the `MusicBrainzDiscIdLoaded` signal, if they need to.

#### Final Note
Let me know if you want me to stop requesting to merge stuff into master - not sure how far along you are with the release candidate...

jbroadus commented on src/devices/cddasongloader.cpp:
```diff
@@ -30,10 +30,19 @@ CddaSongLoader::CddaSongLoader(const QUrl& url, QObject* parent)
+
+  cdio_ = cdio_open(url_.path().toLocal8Bit().constData(), DRIVER_DEVICE);
+  Q_ASSERT(cdio_ != nullptr);  // todo: assert for now, but we need to actually
+                               // handle error cases here
```
Maybe we can add a separate Init() method that returns a bool? CddaDevice already has an Init method that could call it, though it doesn't currently return anything.

lumip replied:
I'm not a fan of Init() methods if they can be avoided. Having an instance of CddaSongLoader that is not in a properly initialised state (i.e. where `Init()` would have returned `false`) seems pointless because it's unusable. Better to fail during construction, imo.

jbroadus commented on src/devices/cddasongloader.cpp:
```diff
@@ -30,10 +30,19 @@ CddaSongLoader::CddaSongLoader(const QUrl& url, QObject* parent)
+  // The LoadSongsFromCdda methods runs concurrently in a thread and we need to
+  // wait for it to terminate. There's no guarantee that it has terminated when
+  // destructor is invoked.
+  loading_future_.waitForFinished();
```
Would there be a way to interrupt LoadSongsFromCdda? There are a couple loops that could check for abort flags.

lumip replied:
Haven't looked into that yet because `LoadSongsFromCdda` shouldn't really take too long to complete in any case, but I can add a flag for it to abort.

lumip replied:
I have now introduced a `may_load_` flag and have `LoadSongsFromCdda` interrupt its metadata parsing loop if that is no longer set.

jbroadus commented on src/devices/cddasongloader.cpp:
```diff
@@ -214,13 +304,9 @@ void CddaSongLoader::AudioCDTagsLoaded(
-    return false;
-  }
-  // Check if mutex is already token (i.e. init is already taking place)
-  if (!mutex_load_.tryLock()) {
```
From cddadevice.cpp:
```
void CddaDevice::Init() {
  song_count_ = 0;  // Reset song count, in case it was already set
  cdda_song_loader_.LoadSongs();
}

void CddaDevice::Refresh() {
  if (!cdda_song_loader_.HasChanged()) {
    return;
  }
  Init();
}
```
So, the mutex is preventing the LoadSongs from being when there's a loading thread running. That synchronization logic really shouldn't be spanning two classes, but it looks like it was the result of CddaDevice being split into two pieces. But if you do want to stick with this existing pattern, then I think you could check the state of loading_future_ here.

The old code also has a timing hole. You could potentially call Refresh multiple times in succession and kick off several threads before the first thread has a chance to grab the lock. I think that could also be fixed by checking the future object.

lumip replied:
I think this could neatly be adressed by my idea of making `CddaSongLoader` more stateful and encapsulating the logic for only re-reading data if the medium was changed. `CddaDevice::Refresh` could then just call `Init` without any checks there

lumip replied:
I have now encapsulated checks for `HasChanged` in  `CddaSongLoader::LoadSongs`. Subsequent calls to `LoadSongs` without a media change will not perform any reading but simply emit a signal with the complete track list with all metadata once. `CddaDevice::Refresh` therefore can just call `::Init()` now and no more synchronisation issues can occur there.

jbroadus commented on src/devices/cddasongloader.cpp:
```diff
@@ -47,18 +69,86 @@ QUrl CddaSongLoader::GetUrlFromTrack(int track_number) const {
 void CddaSongLoader::LoadSongs() {
-  QtConcurrent::run(this, &CddaSongLoader::LoadSongsFromCdda);
+  if (!disc_.has_been_read || HasChanged()) {
+    if (!loading_future_.isRunning()) {
```
There's isStarted and isRunning. I assumed that isStarted() meant that QtConcurrent::run had been called and that isRunning() meant that the thread entry point had been called. But I may be wrong about that. The QFuture documentation isn't terribly helpful.

lumip replied:
Briefly looking at the source of QFuture didn't help me to figure out the difference either. I tried making that condition  `!(loading_future_.isRunning() || loading_future_.isStarted())` to cover both but found that that prevents loading completely. Apparently it holds that `loading_future_.isStarted() == true` after initialisation of `loading_future_` in the class constructor, which seems really odd to me...

jbroadus commented on src/devices/cddasongloader.cpp:
```diff
@@ -47,18 +69,86 @@ QUrl CddaSongLoader::GetUrlFromTrack(int track_number) const {
+          QtConcurrent::run(this, &CddaSongLoader::LoadSongsFromCdda);
+    }
+  } else
+    emit SongsUpdated(disc_.tracks, false);
```
`CddaDevice::Refresh` is called when the view asks for the items display text, though I'm not sure why:
https://github.com/clementine-player/Clementine/blob/master/src/devices/devicemanager.cpp#L194

That means that this signal could be emitted frequently.

lumip replied:
Hm. I'm not sure that's a problem, and if it is, I'd argue it's not one of `CddaSongLoader`. After all, if you ask me to load some songs for you, you shouldn't be surprised when you get a list of songs... I'll see if I can think of a way to make the `CddaDevice::Refresh` smarter (or maybe something even more upstream of that..)

jbroadus replied:
Yeah, the refresh looks like it was a bit of a hack.

This loader is different from the other device loaders (mtp, gpod, etc) in that its lifecycle is the same as the device object. The others are created, preform a single load, and then are deleted. And this is the pattern that SongLoader currently follows when using CddaSongLoader. So, maybe CddaSongLoader should be pared down and all of the synchronization should be handled by the device class instead? I think the cdio handle would need to be passed into CddaSongLoader in that case.

lumip replied:
That's a good point. I was going in the other direction so far (just pushed my pending changes): I have now extended CddaDevice to basically suppress SongsUpdated signals unless there was a substantial change. How this works now is that `CddaSongLoader::LoadSongs` will emit a `DiscChanged` signal if it detects a disc change before it starts reading the disc information. `CddaDevice` will then change an internal flag to accumulate all incoming `SongsUpdated` signals until it encounters one with `futher_updates_possible==false`, upon which it will emit its `SongsLoaded` signal and turn off the internal flag.
For calls to `Refresh` that do not result in a `DiscChanged` signal being received, `CddaDevice` ignores/suppresses the `SongsUpdated` signal it receives from `CddaSongLoader`.

It works now but I'm not really happy with it, because it increases the complexity of the signalling protocol of `CddaSongLoader`. Ideally I'd want that protocol to be stateless, i.e., `SongsUpdated` signals just happen and are self-contained - but that would require some very smart solution to the refreshing issue..

I'm actually thinking of absorbing the media change detection of PR #7024 into `CddaSongLoader` (or `CddaDevice`) and get rid off the `ConnectedDevice::Refresh` method entirely (its only used in `CddaDevice` anyways..). 

Another issue I've now run into is that the `cdio_get_media_changed` seems to not be working reliably when more than one `cdio` object is opened, which is now the case for the cd ripping (once held by `Ripper` and once by `CddaSongLoader`, both of which are used by `RipCDDialog`), which really messes with my changes now (which is another reason that motivates me to centralise all cdio related things in one class).

Thoughts?

jbroadus replied:
I wonder if the cdio instance references a singleton? One solution to that could be a static change counter that increments every time a change is detected and each instance could keep a local copy of the count last time it checked. Or maybe the CddaDevice instance can serve the ripper. I'm not sure how it works right now if you try to rip and play the CD at the same time.

I agree with getting rid of `ConnectedDevice::Refresh`. A timer in CddaDevice or CddaSongLoader makes more sense to me.

Also, doing this in #7024, or even a different PR, sounds like a good idea to me.

lumip replied:
> I wonder if the cdio instance references a singleton?
Looking at the code of the linux driver implementation of cdio, this doesn't seem to be the case. Each call to `cdio_open` creates a new instance and opens a new handle to the device. `cdio_get_media_changed` is just a wrapper around the `CDROM_MEDIA_CHANGED` ioctl call. Therefore, it seems that this might actually be a Linux kernel issue (and might not be an issue on other systems). I have somewhat confirmed this with a small test program that uses the ioctl call directly and could confirm that the call only returns 1 once when invoked for several separate file descriptors - this even holds across several independent processes.
I'll keep your suggestions in mind and try to reduce the number of cdio instances held by Clementine to one.

jbroadus commented on src/devices/cddadevice.cpp:
```diff
@@ -42,17 +38,15 @@ CddaDevice::~CddaDevice() {}
-  }
-  Init();
-}
+void CddaDevice::Refresh() { Init(); }
```
I know this is the existing pattern, but, in my option, it makes more sense to put the code in Refresh() and have Init() call Refresh().

lumip replied:
Done

jbroadus commented on src/devices/cddasongloader.h:
```diff
@@ -42,29 +45,69 @@ class CddaSongLoader : public QObject {
+  // since the last time LoadSongs() was called.
   void LoadSongs();
+  // Returns true if the medium/disc has changed since the last check.
   bool HasChanged();
```
Can this be private now?

lumip replied:
Done

lumip commented:
I have now rebased this to incorporate the changes made to cdio object handling and device access in #7037 .

jbroadus commented on src/devices/cddasongloader.cpp:
```diff
@@ -140,43 +225,70 @@ void CddaSongLoader::LoadSongsFromCdda() {
+  if (!musicbrainz_discid.isEmpty())
+    emit MusicBrainzDiscIdLoaded(musicbrainz_discid);
+  else
+    SongsUpdated(tagged_song_list, false);
```
Should this be an emit?

lumip replied:
True, I have fixed this now.

jbroadus commented on src/devices/cddasongloader.cpp:
```diff
@@ -55,17 +68,88 @@ bool CddaSongLoader::IsActive() const { return loading_future_.isRunning(); }
+}
+
 void CddaSongLoader::LoadSongsFromCdda() {
   if (!may_load_) return;
```
Is the object leaked in this case?

lumip replied:
Which object exactly do you mean? I had a brief look at SongLoader::LoadLocalPartial but can't really see how that interacts with this as it seems concerned with local files, not CDs. Correct me if I'm wrong please.

jbroadus replied:
Sorry, I read the wrong method name. I meant SongLoader::LoadAudioCD(). That creates an instance of CddaSongLoader and relies on a signal to tell it when it can delete the object. We should make sure your finished signal is emitted in all cases.

lumip replied:
Ah, I see. I have added a `Finished` signal to `CddaSongLoader` according to your suggestion below and double checked that every branch that ends the loading process in `CddaSongLoader`, abort or success, emits `Finished`, so I think that the changes work well with `SongLoader::LoadAudioCD`. It is a bit hard to test with 100% certainty, though.

jbroadus commented on src/core/songloader.cpp:
```diff
@@ -156,10 +156,8 @@ SongLoader::Result SongLoader::LoadLocalPartial(const QString& filename) {
-  connect(cdda_song_loader, SIGNAL(SongsMetadataLoaded(SongList)), this,
-          SLOT(AudioCDTracksTagsLoaded(SongList)));
+  connect(cdda_song_loader, SIGNAL(SongsUpdated(SongList, bool)), this,
+          SLOT(AudioCDTracksLoadedSlot(SongList, bool)));
```
While I think that unifying these is mostly correct, I wonder if we should have a separate Finished signal where the connected slot would just delete the object and handle any errors. I think it might be a little clearer than the boolean passed to SongsUpdated.

lumip replied:
That's a good idea and I have changed it accordingly.

jbroadus commented on src/devices/cddasongloader.h:
```diff
@@ -42,30 +43,70 @@ class CddaSongLoader : public QObject {
 
+  // The list of currently cached tracks. This gets updated during calls
+  // LoadSongs() is called. Not thread-safe.
+  SongList chached_tracks() const;
```
Typo

lumip replied:
Indeed. Fixed.

jbroadus commented on src/devices/cddadevice.cpp:
```diff
@@ -73,12 +71,28 @@ void CddaDevice::WatchForDiscChanges(bool watch) {
+void CddaDevice::LoadSongs() {
+  SongList songs = cdda_song_loader_.chached_tracks();
+  SongsLoaded(songs);
+}
```
Maybe there's a better name for this method. It isn't actually loading the tracks anymore.

lumip replied:
True. Any suggestions? Maybe `RequestSongs` or `AnnounceSongs`/`SignalSongs`?

jbroadus replied:
Hmm. Maybe `UpdateFromLoader`?

lumip replied:
But it's also not really updating anything, using just the cached information. The only thing it really does is (re)-emit the signal with the known song list (which gets updated automatically upon media change or on first read). So maybe just `EmitSongs`?

Thinking about it, it seems to indicate that using the signal as the main way of getting the song list to the consumers might not be the best design decision, because it forces this odd re-emission. I could change it so that the `CddaDevice` (and potentially also `CddaSongLoader`) just expose methods to get the song information simply emit a signal whenver that changes. Consumers can then just call the method when receiving the signal. That would also take care of the concern you have expressed in the discussion below, as the signal would not be re-emitted if a new consumer wanted to query the tracks.

jbroadus replied:
Yeah, as long as the fetch is re-entrant, which I think it is.

lumip replied:
I'm not quite sure it is. I have now done the following:
- protected access to the tracklist cache in CddaSongLoader with a mutex
- added method CddaDevice::songs (which forwards to CddaSongLoader::cached_tracks)
- removed CddaDevice::LoadSongs
- renamed the private CddaDevice::ForceLoadSongs back to CddaDevice::LoadSongs
- RipCDDialog now invokes CddaDevice::songs to get the song list from the device instead of provoking it to emit a signal

I didn't change the behavior of the existing signals, so those still provide the updated list of songs to the receiver.

jbroadus commented on src/ripper/ripcddialog.cpp:
```diff
@@ -293,20 +291,12 @@ void RipCDDialog::DeviceSelected(int device_index) {
-          SLOT(AddAlbumMetadataFromMusicBrainz(SongList)));
+  connect(cdda_device_.get(), SIGNAL(SongsDiscovered(SongList)),
+          SLOT(SetAlbumMetadata(SongList)));
+  cdda_device_->LoadSongs();
```
If I'm reading things correctly, the purpose of this is to re-emit the SongsDiscovered signal. Could this have side effects on other users?

jbroadus replied:
Maybe we could keep the song cache in the CddaDevice so that it's always up to date with the model. Then we could just fetch those when this initializes.

lumip replied:
This only re-emits the SongsDiscovered signal if songs were previously already loaded, however, it can also start a genuine loading process if the device wasn't yet opened/loaded. It has side effects on other consumers of the SongsDiscovered signal insofar as there will be a re-emission of that signal. However, the same signal will also happen if the disc was changed and consumers are therefore expected to (and do) treat this as an update to the song list in a reasonable way.


I was thinking of containing the song cache in CddaDevice, but that has the problem that CddaDevice cannot currently detect disc changes, which require an emptying of the cache, because CddaDevice is now the only instance that has access to the cdio object. The whole split between CddaSongLoader and CddaDevice is a bit problematic as they really heavily depend on each other to function properly but I haven't yet found a good solution to that.

jbroadus commented on src/ripper/ripcddialog.cpp:
```diff
@@ -293,20 +291,12 @@ void RipCDDialog::DeviceSelected(int device_index) {
-  connect(loader_, SIGNAL(SongsMetadataLoaded(SongList)),
-          SLOT(AddAlbumMetadataFromMusicBrainz(SongList)));
+  connect(cdda_device_.get(), SIGNAL(SongsDiscovered(SongList)),
+          SLOT(SetAlbumMetadata(SongList)));
```
Would it make sense to connect SongsDiscovered to a single slot?

lumip replied:
We could do that, but I think the logical separation of updating the track list and updating the metadata fields makes some sense as well and helps readability of the code. We could call `BuildTrackListTable` and `SetAlbumMetadata` from inside a new handler for SongsDiscovered, to avoid the multiple connections, though.

lumip commented on src/devices/cddadevice.cpp:
```diff
@@ -73,12 +71,23 @@ void CddaDevice::WatchForDiscChanges(bool watch) {
+  // Ensure that cdio_get_media_changed is cleared
+  // after songs are loaded, so we don't potentially
+  // re-read the same disc.
+  cdio_get_media_changed(cdio_);
```
The problem I was trying to address with this is that cdio_get_media_changed is returns true twice when a disc is inserted, with a bit of delay between both (I think they roughly correspond to "tray is closed" and "disc was detected/read/spun up/whatever"). There's no way to programmatically distinguish them and for the current code this means that the first time cdio_get_media_changed "fires" we start reading the disc, deactivating the disc change polling timer while doing so. After reading is complete, the timer gets reactivated and, due to the behavior described above, sees cdio_get_media_changed returned true, causing another reading of the disc. Therefore I decided to "clear" the cdio_get_media_changed response after reading to avoid this.

I can see your concern with what happens if the media changes during the load process. I tried to provoke this, but I was unable to as the reading is quite fast and I think the tray is locked while that is in progress. I could only get the drive to eject the disc before the read started (while it was still spinning up) or after the read completed (with a bit of a delay for spin down). I think it is near impossible to actually have a disc change in the time between the read completing and this reset of cdio_get_media_changed occuring and it wouldn't be a huge problem anyways (user would just have to cycle the tray once in case it happens).

That said, it just occured to me that I still need to check this without an internet connection since waiting for a response from musicbrainz may delay this long enough for a physical disc change to actually take place...

hatstand commented:
@lumip can you fix the conflicts here?

lumip commented:
I'm working on a rebase to resolve conflicts and nicefy the commits after the edits from the review. Expect that soon.

lumip commented:
@hatstand @jbroadus I took me a bit as I was busy otherwise, but I have now resolved all conflicts and squashed commits somewhat. I've also fixed a bug emerging from what appears to be a subtle race-condition with gstreamer that resulted in CddaSongLoader to skip reading tags for a track randomly in rare cases. Otherwise there are no substantial changes from the last review.
